### PR TITLE
feat(config): configurable bootstrap policies file

### PR DESCRIPTION
**Pull request by sgomezvillamor**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by sgomezvillamor** [comment]
hi @david-leifker! just noted you self-assigned this PR. Some questions before opening for the review:

Do you think some tests should be included for this?
Tested locally with defaut values as well as with env var `BOOTSTRAP_POLICIES_FILE=file:///test/my-policies.json`, all fine! Just not sure how to cover this second scenario in a test 🤔 

Also, should this be documented somewhere?

**Comment by david-leifker** [comment]
I self-assigned because I think that the default policy is loaded once and then never loaded again from the jar. If I recall correctly the state of whether the bootstrap step has run is stored in the SQL database. Overriding this file after the very first run isn't going to work as expected. I wanted to validate that it will behave like that and to ask whether that is your intent (to override this policy once for a new instance)?

**Comment by sgomezvillamor** [comment]
> default policy is loaded once and then never loaded again from the jar. 

https://github.com/datahub-project/datahub/blob/6c55eb8e6d40708849b530be8863035ef9f21a27/metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java#L80-L105

For every policy in the bootstrap file, the following criteria determines whether policy is loaded again or not:
* if is not editable --> reload
* else:
  * if policy not exists in backend --> reload
  * else --> skip

>  I recall correctly the state of whether the bootstrap step has run is stored in the SQL database. 

Haven't found evidence of this. From what I see in the code and logs, `IngestionPoliciesStep` is processed in every boot and the criteria above is followed to determine reload or not.

> Overriding this file after the very first run isn't going to work as expected.

I think it will. If we make the `IngestionPoliciesStep` to process a different file, as soon as we keep stable ids, the criteria above will still be valid.

---

The goal is to use a different file from the one in the jar while keeping all other functionality. There are some policies there we want to disable (even removing) while keeping all policies managed as code. So using a custom bootstrap policies file looks perfect match here.

Please, let me know your thoughts here

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java
Comment (jjoyce0510) [inline metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java:73]: does this correctly print the _path_ of the resource?

or the contents?

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java
Comment (jjoyce0510) [inline metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java:76]: definitely need to ensure that we've confirmed that this works still

**Comment by jjoyce0510** [comment]
Overall I think this looks good. I have not used the dynamic Resource injection via spring but looks like all should work as expected! 

If you provide a file, do you need to prefix the file path similar to how is done for classpath?

e.g. `file:path/to/file`?

If yes, I'd recommend adding a comment in application.yml to exemplify the correct format!

**Comment by jjoyce0510** [comment]
It looks like the syntax indeed would be: 

```
    @Value("file:/path/to/your/file.txt")
```

In this case, Id recommend that we change name of the property inside application.yml to "resource" instead of "file", as file implies a relative path would work!

File: metadata-service/configuration/src/main/resources/application.yml
Comment (sgomezvillamor) [reply metadata-service/configuration/src/main/resources/application.yml:280]: yes, I tested with default values (so loading from classpath) and with `ENV BOOTSTRAP_POLICIES_FILE=file:///datahub/datahub-gms/resources/custom-policies.json`

It works fine in both cases

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java
Comment (sgomezvillamor) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java:73]: The _path_ 

This is the output with default values


```
2023-09-22 10:05:22 2023-09-22 08:05:22,700 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:73 - Ingesting default access policies from: class path resource [boot/policies.json]...
2023-09-22 10:05:22 2023-09-22 08:05:22,710 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:98 - Ingesting default policy with urn urn:li:dataHubPolicy:0
2023-09-22 10:05:22 2023-09-22 08:05:22,725 [main] INFO  c.l.m.entity.EntityServiceImpl:1835 - Ingesting aspect with name dataHubPolicyKey, urn urn:li:dataHubPolicy:0
2023-09-22 10:05:22 2023-09-22 08:05:22,727 [main] INFO  c.l.m.entity.EntityServiceImpl:1835 - Ingesting aspect with name dataHubPolicyInfo, urn urn:li:dataHubPolicy:0
2023-09-22 10:05:22 2023-09-22 08:05:22,746 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:98 - Ingesting default policy with urn urn:li:dataHubPolicy:1
2023-09-22 10:05:22 2023-09-22 08:05:22,762 [main] INFO  c.l.m.entity.EntityServiceImpl:1835 - Ingesting aspect with name dataHubPolicyKey, urn urn:li:dataHubPolicy:1
2023-09-22 10:05:22 2023-09-22 08:05:22,762 [main] INFO  c.l.m.entity.EntityServiceImpl:1835 - Ingesting aspect with name dataHubPolicyInfo, urn urn:li:dataHubPolicy:1
2023-09-22 10:05:22 2023-09-22 08:05:22,810 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:106 - Skipping ingestion of editable policy with urn urn:li:dataHubPolicy:7
2023-09-22 10:05:22 2023-09-22 08:05:22,831 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:106 - Skipping ingestion of editable policy with urn urn:li:dataHubPolicy:view-entity-page-all
2023-09-22 10:05:22 2023-09-22 08:05:22,841 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:106 - Skipping ingestion of editable policy with urn urn:li:dataHubPolicy:view-dataset-sensitive
2023-09-22 10:05:22 2023-09-22 08:05:22,847 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:98 - Ingesting default policy with urn urn:li:dataHubPolicy:admin-platform-policy
2023-09-22 10:05:22 2023-09-22 08:05:22,863 [main] INFO  c.l.m.entity.EntityServiceImpl:1835 - Ingesting aspect with name dataHubPolicyKey, urn urn:li:dataHubPolicy:admin-platform-policy
...
```

And this is the output with `ENV BOOTSTRAP_POLICIES_FILE=file:///datahub/datahub-gms/resources/custom-policies.json`

```

2023-09-22 10:45:27 2023-09-22 08:45:27,027 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:73 - Ingesting default access policies from: URL [file:/datahub/datahub-gms/resources/custom-policies.json]...
2023-09-22 10:45:27 2023-09-22 08:45:27,032 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:98 - Ingesting default policy with urn urn:li:dataHubPolicy:XXX
2023-09-22 10:45:27 2023-09-22 08:45:27,132 [main] WARN  org.apache.velocity.deprecation:62 - configuration key 'resource.loader' has been deprecated in favor of 'resource.loaders'
2023-09-22 10:45:27 2023-09-22 08:45:27,133 [main] WARN  org.apache.velocity.deprecation:62 - configuration key 'class.resource.loader.class' has been deprecated in favor of 'resource.loader.class.class'
2023-09-22 10:45:27 2023-09-22 08:45:27,133 [main] WARN  org.apache.velocity.deprecation:62 - configuration key 'file.resource.loader.class' has been deprecated in favor of 'resource.loader.file.class'
2023-09-22 10:45:27 2023-09-22 08:45:27,133 [main] WARN  org.apache.velocity.deprecation:62 - configuration key 'file.resource.loader.path' has been deprecated in favor of 'resource.loader.file.path'
2023-09-22 10:45:27 2023-09-22 08:45:27,133 [main] WARN  org.apache.velocity.deprecation:62 - configuration key 'runtime.references.strict' has been deprecated in favor of 'runtime.strict_mode.enable'
2023-09-22 10:45:27 2023-09-22 08:45:27,627 [main] INFO  c.l.m.boot.steps.IngestPoliciesStep:115 - Successfully ingested default access policies.
```

File: metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java
Comment (sgomezvillamor) [reply metadata-service/factories/src/main/java/com/linkedin/metadata/boot/steps/IngestPoliciesStep.java:76]: It does, see comment https://github.com/datahub-project/datahub/pull/8812/files#r1334091670

**Comment by sgomezvillamor** [comment]
> Overall I think this looks good. I have not used the dynamic Resource injection via spring but looks like all should work as expected!

I also learnt that in this contrib 😅  https://www.baeldung.com/spring-load-resource-as-string#using-spel


> Overall I think this looks good. I have not used the dynamic Resource injection via spring but looks like all should work as expected!
> 
> If you provide a file, do you need to prefix the file path similar to how is done for classpath?
> 
> e.g. file:path/to/file?
>
> If yes, I'd recommend adding a comment in application.yml to exemplify the correct format!

With `file:...` definitely works. I tested with `ENV BOOTSTRAP_POLICIES_FILE=file:///datahub/datahub-gms/resources/custom-policies.json` in my local deployment. Not sure if there are other variants that could work also.
I will add the comment/example in `application.yml` 👍

**Comment by jjoyce0510** [comment]
LGTM- Once we are green we merge!

**Review by jjoyce0510 — APPROVED** [review]

**Pull request merged**
